src/ApolloHooksMutation.re

```diff
@@ -0,0 +1,210 @@
+type graphqlErrors;
+
+type error = {
+  .
+  "message": string,
+  "graphqlErrors": graphqlErrors,
+};
+
+type refetchQueries =
+  ReasonApolloTypes.executionResult => array(ApolloClient.queryObj);
+
+/* Result that is return by the hook */
+type result('a) =
+  | Data('a)
+  | Error(error)
+  | NoData;
+
+/* Result that is return by the hook */
+type controlledResult('a) = {
+  loading: bool,
+  called: bool,
+  data: option('a),
+  error: option(error),
+};
+
+type controlledVariantResult('a) =
+  | Loading
+  | Called
+  | Data('a)
+  | Error(error)
+  | NoData;
+
+[@bs.module "graphql-tag"] external gql: ReasonApolloTypes.gql = "default";
+
+type mutationResult('a) = {. "data": 'a};
+
+[@bs.deriving abstract]
+type options('a) = {
+  [@bs.optional]
+  variables: Js.Json.t,
+  [@bs.optional]
+  client: ApolloClient.generatedApolloClient,
+  [@bs.optional]
+  refetchQueries,
+  [@bs.optional]
+  awaitRefetchQueries: bool,
+  [@bs.optional]
+  update: (ApolloClient.generatedApolloClient, mutationResult('a)) => unit,
+};
+
+type jsResult = {
+  .
+  "data": Js.Nullable.t(Js.Json.t),
+  "loading": bool,
+  "called": bool,
+  "error": Js.Nullable.t(error),
+};
+
+type jsMutate('a) = (. options('a)) => Js.Promise.t(jsResult);
+type mutation('a, 'b) =
+  (
+    ~mutation: 'b,
+    ~client: ApolloClient.generatedApolloClient=?,
+    ~refetchQueries: refetchQueries=?,
+    ~awaitRefetchQueries: bool=?,
+    unit
+  ) =>
+  Js.Promise.t(controlledVariantResult('a));
+
+[@bs.module "@apollo/react-hooks"]
+external useMutationJs:
+  (. ReasonApolloTypes.queryString, options('a)) => (jsMutate('a), jsResult) =
+  "useMutation";
+
+exception Error(string);
+
+let useMutation =
+    (
+      ~incompleteMutation=?,
+      ~mutation=?,
+      ~client=?,
+      ~refetchQueries=?,
+      ~awaitRefetchQueries=?,
+      ~update=?,
+      (),
+    ) => {
+  let (jsMutate, jsResult) =
+    useMutationJs(.
+      gql(.
+        switch (mutation, incompleteMutation) {
+        | (Some(mutation), _) => mutation##query
+        | (_, Some(incompleteMutation)) => incompleteMutation
+        | (None, None) =>
+          raise(
+            Error("You need to provide a mutation or an incomplete mutation"),
```

Comment:
I think we need to have a better exception type, something more descriptive than `Error`, does it makes sense, @fakenickels ?

Reply:
Maybe we can avoid throwing errors here by making sure the user can't make a mistake on the type level? I think it will be better to not throw errors if the type system can guarantee that the values are correct 😄 I was thinking about using some kind of a variant:
```
type mutationConfig('a) = Query(string) | Config('a);
```
and then have it in `useMutation`:
```
let useMutation =
    (
      ~mutation: mutationConfig('a)=?,
      ~client=?,
     ....
      (),
    )
```

Reply:
Ideally we can remove the need to pass in incompleteMutation in the future which will solve this issue.